### PR TITLE
RSDK-8819: Instrument robot with FTDC. Gated on startup flag.

### PR DESCRIPTION
**Pull request by dgottlieb**
Another mess of a PR. Annotating inline

### CODE REVIEW

File: ftdc/custom_format.go
Comment (dgottlieb) [inline ftdc/custom_format.go:35]: The following changes are me bubbling errors back up from failure to write data.

File: ftdc/custom_format.go
Comment (dgottlieb) [inline ftdc/custom_format.go:69]: Going through this function, I added/edited some documentation.

File: ftdc/custom_format_test.go
Comment (dgottlieb) [inline ftdc/custom_format_test.go:281]: I guess I can add the assertions now! oops

File: ftdc/ftdc.go
Comment (dgottlieb) [inline ftdc/ftdc.go:99]: So my first use of the ticket case I added a bit ago was....one ticker worker (reading stats) and one non-ticker worker (writing stats).

Which we explicitly didn't set out to support. But now I have an idea of the requirements. The non-ticker worker is consuming from a channel written to by the ticker worker.

File: ftdc/ftdc.go
Comment (dgottlieb) [inline ftdc/ftdc.go:129]: Oops, can remove

File: ftdc/ftdc.go
Comment (dgottlieb) [inline ftdc/ftdc.go:213]: The business logic for errors in the write-path. I think all of the panics I removed get to this line of code now.

File: resource/graph_node.go
Comment (dgottlieb) [inline resource/graph_node.go:598]: I can experiment with removing this again. This was part of the whole "oh crap I didn't support nested `any`".

File: robot/impl/resource_manager.go
Comment (dgottlieb) [inline robot/impl/resource_manager.go:151]: I was not happy with my options for these swap resource usages. So I just left in the dumbest/most straightforward code. Rather than spending time trying to come up with something clever on my own.

This is a pretty antagonistic case. I need to document why I did this. But now that I see I omitted the documentation, I'm curious if you can figure out what this protects against.

File: robot/impl/robot_options.go
Comment (dgottlieb) [inline robot/impl/robot_options.go:22]: I created a `./viam-server -ftdc` flag for this and threaded it through. As mentioned at the prior meeting, can talk with the team if we we want to softly gate this to start.

File: ftdc/custom_format.go
Comment (benjirewis) [reply ftdc/custom_format.go:138]: What happens if `ResourceStats` is an interface besides `any`? Or are you saying that for any interface type you have to "flatten" with `inp = inp.Elem()`?

File: ftdc/custom_format.go
Comment (benjirewis) [inline ftdc/custom_format.go:149]: What is this predicate saying? If a flattened pointer is still a pointer or interface and that pointer is nil then return `[]float32{}`? Makes sense to me, but double-checking my understanding.

File: ftdc/custom_format.go
Comment (benjirewis) [reply ftdc/custom_format.go:204]: > I'm also concerned (FUD until proven otherwise) that merging them will make the "steady state" path of just caring about float32 values slower.

I don't understand this piece, but not merging for now despite the common `reflect.Value` param seems fine to me.

File: ftdc/custom_format.go
Comment (benjirewis) [inline ftdc/custom_format.go:201]: > does not recursively walk data structures

No longer accurate? Or am I misunderstanding?

File: ftdc/ftdc.go
Comment (benjirewis) [reply ftdc/ftdc.go:100]: IIRC as long as we don't have > 1 ticker worker, you should be good.

File: ftdc/ftdc.go
Comment (benjirewis) [inline ftdc/ftdc.go:179]: Don't want a `PanicCapturingGo`?

File: ftdc/ftdc.go
Comment (benjirewis) [reply ftdc/ftdc.go:213]: Any `writeDatum` error is fatal for FTDC?

File: resource/graph_node.go
Comment (benjirewis) [inline resource/graph_node.go:575]: Why not `NodeState`?

File: robot/impl/resource_manager.go
Comment (benjirewis) [reply robot/impl/resource_manager.go:151]: Yeah I wouldn't have expected needing `Remove` and `Add` surrounding `SwapResource`. I'm assuming FTDC was not picking up on new `Stats` implementations for reconfigured resources? No clue, in any case, some documentation here would be great.

File: robot/impl/resource_manager.go
Comment (benjirewis) [inline robot/impl/resource_manager.go:154]: Uh lost again here; when would that happen?

File: robot/impl/robot_options.go
Comment (benjirewis) [reply robot/impl/robot_options.go:22]: Feels like a fine way to gate (cc @viamrobotics/netcode if others have strong opinions).

File: web/server/entrypoint.go
Comment (benjirewis) [inline web/server/entrypoint.go:51]:
```suggestion
	EnableFTDC                 bool   `flag:"ftdc,usage=enable fulltime data capture for diagnostics [beta feature]"`
```

File: ftdc/custom_format_test.go
Comment (benjirewis) [reply ftdc/custom_format_test.go:281]: Yeah assertions here would be great.

**Review by benjirewis — COMMENTED** [review]
Looking good! Lots of qs.

File: ftdc/custom_format.go
Comment (dgottlieb) [reply ftdc/custom_format.go:138]: > Or are you saying that for any interface type you have to "flatten" with inp = inp.Elem()?

Yeah, that.

File: ftdc/custom_format.go
Comment (dgottlieb) [reply ftdc/custom_format.go:204]: I'm assuming you're curious why I think it would be less performant.

I think to merge the functions we'd need something like this:
```
type NameValue struct{
  name string
  value float32
}

func getFieldNamesAndValuesForStruct(item reflect.Value) []NameValue {
...
}
```

And now when I get the schema, I'm creating a bunch of strings (necessary) and a bunch of floats (unnecessary). That part I'm not too worried about. I worry more about the "steady state" where the schema doesn't change and we're just recording new values.

In that case we create a bunch of strings (unnecessary) and a bunch of floats (necessary). I worry that all of the extra allocations, just to be thrown away will "quickly" add up.

File: ftdc/ftdc.go
Comment (dgottlieb) [reply ftdc/ftdc.go:100]: I think we can add it, but I'm not positive that means I can remove the `outputWorkerDone chan struct{}` variable (which is a non-trivial line of reasoning -- the margin is too small for this proof).

And if I can't use the StoppableWorkers to _completely_ control shutdown, I'd prefer to stay on track of getting FTDC deployed.

FWIW, I think we can add to StoppableWorkers to make this possible, but I've already done enough speculative patch pushing that was only 90% usable.

File: ftdc/ftdc.go
Comment (dgottlieb) [reply ftdc/ftdc.go:213]: Oh shoot, good catch (or at least I think* you caught it).

The intention was that errors bubbling up to here are all disk related. And I'm not yet interested in making this robust across disk errors.

But one of the errors that can bubble up is the schema error. That we technically handle at a lower level (that I think is a reasonable spot). And I think bubbling it up is reasonable? But we certainly shouldn't shutdown ftdc over that case.

File: robot/impl/resource_manager.go
Comment (dgottlieb) [reply robot/impl/resource_manager.go:151]: > I'm assuming FTDC was not picking up on new Stats implementations for reconfigured resources?

I didn't observe this, I was just checking my assumptions on how schemas can change w.r.t underlying resources. But what you said is correct. A model can change causing a SwapResource. Which means a different object we're calling `Stats` on.

> No clue, in any case, some documentation here would be great.

Agreed, that's a good idea.

Comment (dgottlieb) [reply robot/impl/resource_manager.go:151]: Actually, the slightly more clever thing is to add a `SwapResourceWithFTDCTracking`.

Or I could always make SwapResources take an FTDC, but I'm feeling a bit lazy atm.

File: ftdc/custom_format.go
Comment (dgottlieb) [reply ftdc/custom_format.go:201]: No longer accurate, updated. Thanks for spotting!

File: ftdc/ftdc.go
Comment (dgottlieb) [reply ftdc/ftdc.go:179]: Added

File: resource/graph_node.go
Comment (dgottlieb) [reply resource/graph_node.go:575]: Didn't think about this. But looking at the ultimate output, I think `State` is suitable. I agree the state is tracked by the "graph node", but it's really representing the resource. Here's a screenshot of how the "system name" + field name looks in FTDC graph form (not to be confused with "resource graph" and "graph node").

image

Which reminds me I need to remember to escape underscores! Forgot to port that from the python I wrote:
```
            title = key.replace('_', '\\_')
```

File: robot/impl/resource_manager.go
Comment (dgottlieb) [reply robot/impl/resource_manager.go:154]: This was just me reasoning about the code. We're in `addRemote`, but there's this path about replacing. So I assumed we could go from this original config:
```
{
  components: [{name: "fakeMotor", type: "motor", model: "fake"}],
  remotes: []
}
```

to this interesting (read: antagonistic) config:
```
{
  components: [],
  remotes: [{name: "fakeMotor", address: "..."}]
}
```

Comment (dgottlieb) [reply robot/impl/resource_manager.go:154]: But maybe the name encodes the API + type so it'd be impossible to craft a "name conflict" in this way that confuses the resource manager/graph?

**Comment by dgottlieb** [comment]
It seems my prior patch CI run failed one of my tests:
image

And without knowing that, the questions you asked made me revisit the reflection code that wound up squashing that bug. Without me even knowing that bug existed.

* Glad I had a test that I broke something
* You asked very good/tasteful probing questions (found the right spots that were worth focusing on)

File: ftdc/custom_format.go
Comment (benjirewis) [reply ftdc/custom_format.go:214]: Ah that makes sense to me; didn't immediately understand "steady state". Thanks for more context. Again, leaving as separate for now seems good.

File: ftdc/ftdc.go
Comment (benjirewis) [reply ftdc/ftdc.go:101]: I would buy that `StoppableWorkers` is not enough here and you specifically need synchronization in the form of a done channel for the output worker. What you have seems good, and I wouldn't want to "force" the other concurrency primitives into a `StoppableWorkers`.

File: ftdc/ftdc.go
Comment (benjirewis) [reply ftdc/ftdc.go:220]: I didn't really catch much, but thanks for verifying that not _all_ `writeDatum` errors are fatal to FTDC. I assume `errNotStruct` will get warn or error-logged at some point deeper in the code?

File: resource/graph_node.go
Comment (benjirewis) [inline resource/graph_node.go:260]: Yeah this is much easier to reason about; thanks!

File: robot/impl/resource_manager.go
Comment (benjirewis) [reply robot/impl/resource_manager.go]: > But maybe the name encodes the API + type so it'd be impossible to craft a "name conflict" in this way that confuses the resource manager/graph?

That is my understanding, yes. I don't think you could accidentally `SwapResource` and, in doing so, switch APIs because, as you mention, the resource name encodes the API + type.

File: resource/graph_node.go
Comment (benjirewis) [reply resource/graph_node.go:593]: Let's discuss offline, but that graph is confusing to me.

Comment (benjirewis) [reply resource/graph_node.go:593]: Discussed offline; didn't understand that line on the top right was a legend 😮‍💨 .

I'm still a little nervous about now having two "states" for a resource, one being `NodeState` and one being the reported state you have below based on the error returning from `w.Resource()`. Why would your states based on `err` be better than `NodeState`?

File: ftdc/ftdc.go
Comment (dgottlieb) [reply ftdc/ftdc.go:220]: That was the idea -- I had meant to add a test (but forgot going through the other comments). But I think I still did this wrong

Comment (dgottlieb) [reply ftdc/ftdc.go:221]: Ok, I did a thing. Not 100% I'm making the best use of error handling tools available.

But I do have a test that I can make pass/fail as I break the code. So I'm confident this does what I want.

File: resource/graph_node.go
Comment (dgottlieb) [reply resource/graph_node.go:588]: Despite having a formal "node state" value, whether a resource is accessible (`func graphNode.Resource() (Resource, error)`) to me is the definitive "can a user call methods on this".

The [resource method](https://github.com/viamrobotics/rdk/blob/85f5b08edea7f65d27668359455dfed8f54dad73/resource/graph_node.go?plain=1#L143-L158), for better or worse, as implemented right now does not exhaust the node states. It only checks it for "removing".

I think if we're going to track one thing -- the output of "Resource" is the only thing that matters. If we want to supplement that with a second, more direct mapping of the `state NodeState` [value](https://github.com/viamrobotics/rdk/blob/85f5b08edea7f65d27668359455dfed8f54dad73/resource/graph_node.go?plain=1#L72), we can? I'd prefer to avoid tracking [each variable directly](https://github.com/viamrobotics/rdk/blob/85f5b08edea7f65d27668359455dfed8f54dad73/resource/graph_node.go?plain=1#L58-L67). But that's a straw-man argument, I know you're not actually asking for that.

File: resource/graph_node.go
Comment (dgottlieb) [reply resource/graph_node.go:613]: I took this out, I think `nil` is better.

File: ftdc/ftdc.go
Comment (benjirewis) [reply ftdc/ftdc.go:221]: LGTM! Especially if test indicates you're covering "catch" path.

File: resource/graph_node.go
Comment (benjirewis) [reply resource/graph_node.go:588]: I buy the "can a user call methods on this" argument, but that just makes me think that the formal "node state" is not adequately representative of the state of the resource. I also agree that we shouldn't be tracking both the state you have here and `NodeState` (nor other fields on the graph node). 

I suppose I'm fine with what you have now, but I feel that `NodeState` probably needs to be representative of "can a user call methods on this" (cc @maximpertsov). Once that happens (is it not feasible for some reason I'm not understanding?), then I'd definitely prefer the FTDC-recorded state to be that `NodeState`.

**Review by benjirewis — COMMENTED** [review]
Pretty much LGTM; just want a resolution on state vs `NodeState`.

File: resource/graph_node.go
Comment (dgottlieb) [reply resource/graph_node.go:588]: I'm on board with "as `.Resource()` changes, `Stats` shall too change".

Also I'm on board with just changing the names/values in the future if evidence suggests something else is better.

**Review by benjirewis — APPROVED** [review]
LGTM!

File: robot/impl/robot_options.go
Comment (maximpertsov) [reply robot/impl/robot_options.go:22]: @dgottlieb sorry for the late comment. this gating approach seems fine to me. 

my one question is, do we anticipate that "ftdc = true" will one day be the default? if so, I think it would be bit more ergonomic to transition this to a "noftdc" flag or something like that with a default of false vs having this existing flag default to true. i think that would play more nicely with how golang sets implicit defaults.

anyway, just a thought - no request to change anything!

**Pull request merged**
